src/dctkit/dec/cochain.py

```diff
@@ -394,3 +394,76 @@ def sym(c: Cochain) -> Cochain:
         its symmetric part.
     """
     return scalar_mul(add(c, transpose(c)), 0.5)
+
+
+def convolution(c: Cochain, kernel: Cochain, kernel_window: float) -> Cochain:
+    """ Compute the convolution between two scalar 0-cochains.
+
+    Args:
+        c: a scalar 0-cochain.
+        kernel: the scalar 0-cochain kernel.
+        kernel_window: the kernel window.
+
+    Returns:
+        the convolution rho*kernel.
+    """
+    # we build a kernel matrix K by rolling the kernel vector k + 1 times, where
+    # k is the kernel window. In this way we can express the
+    # convolution between c and k as SK @c_coeffs where S is the hodge star
+    n = len(c.coeffs)
+    K = jnp.zeros((n, n), dtype=dt.float_dtype)
+    buffer = jnp.empty((n, n*2 - 1))
+
+    # generate a wider array that we want a slice into
+    buffer = buffer.at[:, :n].set(kernel.coeffs[:n].T)
+    buffer = buffer.at[:, n:].set(kernel.coeffs[:n-1].T)
+
+    rolled = buffer.reshape(-1)[n-1:-1].reshape(n, -1)
+    K_full_roll = jnp.roll(rolled[:, :n], shift=1, axis=0)
+    K_non_zero = K_full_roll[:n - kernel_window + 1]
+    K = K.at[:n - kernel_window + 1, :].set(K_non_zero)
+    kernel_coch = Cochain(c.dim, c.is_primal, c.complex, K)
+
+    # apply hodge star
+    star_kernel = star(kernel_coch)
+    conv = Cochain(c.dim, c.is_primal, c.complex, star_kernel.coeffs@c.coeffs)
+    return conv
+
+
+def constant_sub(k: float, c: Cochain) -> Cochain:
```

Comment:
I don't like this. A constant cochain is not special, we don't need a specific operation for this. If you need a wrapper, use it in the code of the specific application you are developing, not as a part of the dctkit library.

Reply:
Removed.